sdk/python/feast/job_service.py

```diff
@@ -145,6 +170,78 @@ def GetJob(self, request, context):
         return GetJobResponse(job=self._job_to_proto(job))
 
 
+def get_expected_job_hash_to_table_refs(client) -> Dict[str, Tuple[str, str]]:
+    """Get the map of job hash -> (project, table_name) for expected stream ingestion jobs."""
+    job_hash_to_table_refs = {}
+
+    for project in client.list_projects():
+        feature_tables = client.list_feature_tables(project)
+        for feature_table in feature_tables:
+            if feature_table.stream_source is not None:
+                params = get_stream_to_online_ingestion_params(
+                    client, project, feature_table, []
+                )
+                job_hash = params.get_job_hash()
+                job_hash_to_table_refs[job_hash] = (project, feature_table.name)
+
+    return job_hash_to_table_refs
+
+
+def ensure_stream_ingestion_jobs(client: feast.Client):
```

Comment:
Please add docstrings to methods with types.

Reply:
It really seems like this `ensure` function should be a method on the client itself and not specific to the job service.

Reply:
Can you elaborate, why? It seems like this is a specific functionality to Job Service. Why should client worry about this? Also, this'll make it difficult to decouple client & job service.

Reply:
`ensure` is the same as `start all jobs necessary to be started`. If we believe that starting a specific job is important (from client) then ensure is also important. In fact there is already a use case for this when you cant run stateful processes but you can launch jobs (Databricks scheduler or Github Actions launching Spark jobs on Databricks/EMR).

> Also, this'll make it difficult to decouple client & job service.
How so? Wouldn't it improve the decoupling if the logic is only in the client?

Reply:
tl;dr: discussed in Zoom, see above comment.